spanner/cloudexecutor/executor/cloud_executor_util.go

```diff
@@ -0,0 +1,332 @@
+		return nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "there is no metadata for table %s", tableName))
+	}
+	colMetadata, ok := cols[colName]
+	if !ok {
+		log.Printf("Metadata for table %s contains no column named %s", tableName, colName)
+		return nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "metadata for table %s contains no column named %s", tableName, colName))
+	}
+	return colMetadata.GetType(), nil
+}
+
+// getColumnTypes returns a list of column types of the given table.
+func (t *tableMetadataHelper) getColumnTypes(tableName string) ([]*spannerpb.Type, error) {
+	cols, ok := t.tableColumnsInOrder[tableName]
+	if !ok {
+		log.Printf("There is no metadata for table %s. Make sure that StartTransactionAction has TableMetadata correctly populated.", tableName)
+		return nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "there is no metadata for table %s", tableName))
+	}
+	var colTypes []*spannerpb.Type
+	for _, col := range cols {
+		colTypes = append(colTypes, col.GetType())
+	}
+	return colTypes, nil
+}
+
+// getKeyColumnTypes returns a list of key column types of the given table.
+func (t *tableMetadataHelper) getKeyColumnTypes(tableName string) ([]*spannerpb.Type, error) {
+	cols, ok := t.tableKeyColumnsInOrder[tableName]
+	if !ok {
+		log.Printf("There is no metadata for table %s. Make sure that StartTxnAction has TableMetadata correctly populated.", tableName)
+		return nil, fmt.Errorf("there is no metadata for table %s", tableName)
+	}
+	var colTypes []*spannerpb.Type
+	for _, col := range cols {
+		colTypes = append(colTypes, col.GetType())
+	}
+	return colTypes, nil
+}
+
+// outcomeSender is a utility class used for sending action outcomes back to the client. For read
+// actions, it buffers rows and sends partial read results in batches.
+type outcomeSender struct {
+	actionID int32
+	stream   executorpb.SpannerExecutorProxy_ExecuteActionAsyncServer
+
+	// partialOutcome accumulates rows and other relevant information
+	partialOutcome *executorpb.SpannerActionOutcome
+	readResult     *executorpb.ReadResult
+	queryResult    *executorpb.QueryResult
+
+	// All the relevant variables below should be set before first outcome is sent back,
+	// and unused variables should leave null.
+	timestamp              *timestamppb.Timestamp
+	hasReadResult          bool
+	hasQueryResult         bool
+	hasChangeStreamRecords bool
+	table                  string  // name of the table being read
+	index                  *string // name of the secondary index used for read
+	requestIndex           *int32  // request index (for multireads)
+	rowType                *spannerpb.StructType
+
+	// Current row count in read/query result
+	rowCount int64
+	// modified row count in dml result
+	rowsModified []int64
+}
+
+// if rowCount exceed this value, we should send rows back in batch.
+const maxRowsPerBatch = 100
```

Comment:
Does go-lang follow any style signatures of defining all constant at the start? I would have expected this to be present before any of the other method definitions.